### PR TITLE
implement BridgeStan download and module compilation on Rust

### PR DESCRIPTION
**Pull request by randommm**
Implement BridgeStan download and module compilation on Rust.

Fixes issue #100

### CODE REVIEW

File: rust/src/download_compile.rs
Comment (WardBrian) [inline rust/src/download_compile.rs:66]: It would be nice to pass extra args like we do in the other language. In particular, adding `STANCFLAGS+=--include-paths=.` is needed for a lot of stan code "in the wild"

File: rust/tests/model.rs
Comment (WardBrian) [inline rust/tests/model.rs:1]: Would you mind adding some tests specific to the new functionality?

**Review by WardBrian — CHANGES_REQUESTED** [review]
Thank you for taking this on! A few thoughts on what is here so far

**Comment by randommm** [comment]
> Thank you for taking this on! A few thoughts on what is here so far

okay, all done

**Comment by WardBrian** [comment]
You may have to skip some of the compile tests on Windows for Rust, since we never unload libraries (#111) and Windows locks the file of loaded libraries

**Comment by randommm** [comment]
i see, what about Mac? it seems to give some weird error that is hard to understand.

**Comment by randommm** [comment]
there also seems to be some toolchain problem on Windows, lack of mingw, so it's unable to compile the example.

File: rust/README.md
Comment (WardBrian) [inline rust/README.md:18]: I think this and the later note about how an error will be thrown at runtime if it is not are still worth saying, if a user is using a library compiled from outside of their Rust program.

File: rust/.vscode/settings.json
Comment (WardBrian) [inline rust/.vscode/settings.json:1]: Could you remove this file?

File: .github/workflows/main.yaml
Comment (WardBrian) [inline .github/workflows/main.yaml:293]: I'm a bit concerned that --test-threads=1 will have some impact on our other tests, but I don't recall entirely what it means. I think @aseyboldt and I discussed this when the rust bindings were initially developed

File: rust/src/download_compile.rs
Comment (WardBrian) [inline rust/src/download_compile.rs:117]: I suspect this is why it was failing on Windows - the make command needed there is `mingw32-make`, not `make`

File: rust/src/bs_safe.rs
Comment (WardBrian) [inline rust/src/bs_safe.rs:106]: This is a question because I'm still relatively new to Rust: is putting specific parts of an enum behind a config a fine thing to do?

I guess the enum is already `non_exhaustive`, so it seems fine, but wanted to ask

File: rust/src/download_compile.rs
Comment (WardBrian) [inline rust/src/download_compile.rs:26]: I know we do it in the other languages, but for some reason printing here feels less appropriate than in the scripting languages. OTOH, we do want to let people know what we're doing.

**Review by WardBrian — CHANGES_REQUESTED** [review]
Thanks again. A few more comments below.

I'm also going to ask @aseyboldt if he would be willing to take a look at this

File: rust/README.md
Comment (randommm) [reply rust/README.md:18]: Fair point, the reason i removed it is that if the model compile method is invoked, then it will recompile the model if it wasn't compiled with stan threads previously, but it's true that there is always a risk of someone not calling the compile method

File: .github/workflows/main.yaml
Comment (randommm) [reply .github/workflows/main.yaml:293]: This is the number of threads that are ran in parallel, the reason I added this is because there is a race condition on the compile test because it deletes the .so file before proceding with the compilation and this can interfere with the other tests.

Comment (randommm) [reply .github/workflows/main.yaml]: as an alternative i've added a "ignore" flag to this new test case (on the rust file), so then `cargo test` won't run it. So, then added a new line to main.yaml to run just this test after the other ones.

File: rust/src/download_compile.rs
Comment (randommm) [reply rust/src/download_compile.rs]: I think you're right, I changed the code to use log::info instead of println

File: rust/src/bs_safe.rs
Comment (randommm) [reply rust/src/bs_safe.rs:106]: tricky question, according to my reading of this discussion https://www.reddit.com/r/rust/comments/xyik51/is_feature_gating_some_varaints_of_an_enum_a_bad/ it is acceptable if the enum is non_exhaustive, it's a controversial topic like, e.g. some people are against the use of box leak no matter what.

File: rust/Cargo.toml
Comment (aseyboldt) [inline rust/Cargo.toml:22]: I think we should not expose the optional dependencies as features, ie use the "dep:cratename" syntax mentioned here: https://doc.rust-lang.org/cargo/reference/features.html#optional-dependencies

File: rust/src/download_compile.rs
Comment (aseyboldt) [inline rust/src/download_compile.rs:88]: Why not https://doc.rust-lang.org/std/fs/fn.canonicalize.html
Couldn't we drop the path_absolutize dep then?

File: rust/src/download_compile.rs
Comment (aseyboldt) [inline rust/src/download_compile.rs:103]: Maybe use https://doc.rust-lang.org/std/env/consts/constant.DLL_EXTENSION.html

File: rust/src/download_compile.rs
Comment (aseyboldt) [inline rust/src/download_compile.rs:105]: Why do we need the include-paths here?

File: rust/src/download_compile.rs
Comment (aseyboldt) [inline rust/src/download_compile.rs:129]: Maybe move to after the error check?

File: .github/workflows/main.yaml
Comment (aseyboldt) [inline .github/workflows/main.yaml:294]: I think we should test without the compile feature as well, just to make sure the `if cfg` are all correct.

**Review by aseyboldt — CHANGES_REQUESTED** [review]
Sorry for the late reply...
@randommm Thank you for tackling this, I think I could use this in a couple of places myself.

I marked a couple relatively minor issues that I think should be pretty easy to fix.

I think I'd be fine with the PR if they are addressed, but there are two more other things that I'd also love to have, but they might be a bit more involved:

- Right now it is not obvious for a user of the API that some additional software is being downloaded. This has security implications, as that software could do anything it wants in theory. Maybe we could change the API a bit so that this is at least less hidden?
Maybe we also want to check the downloaded stan version somehow? If we tie it to a specific release we could just use a checksum I guess? Although I'm actually not sure how we'd best do that given that the file we want to download contains the repo, which would have to include the checksum...
- The model compilation function doesn't really work properly if used concurrently, even though nothing on the API side indicates that this is the case. I guess strictly speaking this probably isn't against the safety guarantees of rust, but it still seems to be against the spirit of those (I guess this is why you had to disable the test, or disable testing in parallel). Unfortunately this doesn't seem to be trivial to fix. We could just have a global Mutex that protects the compilation, but that would still fail if multiple independent processes try to compile models in parallel? (@WardBrian Is this something that has been looked into on the stan side?) Maybe we could use some locking mechanism to avoid this? So for instance we could add a `compiler_lock` file in the stan repository, and use fslock on that file to prevent other processes from using the stan installation at the same time. And then we could expose that on the rust side using something like:

```rust
bridgestan::Compiler::new()
    .ensure_download()?
    .lock()  // calls fslock on the lockfile inside the downloaded code
    .comile_model(path_to_model)?
// drop of the lock guard unlocks the lockfile
```

File: rust/src/download_compile.rs
Comment (randommm) [reply rust/src/download_compile.rs:88]: this has some terrible "bug" on windows https://github.com/rust-lang/rust/issues/42869 It was hitting the CI/CD a few commits ago if you wish to check. I think std::path::absolute https://doc.rust-lang.org/std/path/fn.absolute.html would work here but is nightly.

File: rust/src/download_compile.rs
Comment (randommm) [reply rust/src/download_compile.rs:103]: I thought about this too, but it seems that the Python interface always create a file .so extension, so i imagined this is intended?

File: rust/src/download_compile.rs
Comment (randommm) [reply rust/src/download_compile.rs:105]: was included on Python binding so i thought it was usefull somehow https://github.com/roualdes/bridgestan/blob/b41f83b61571d4744ba5521a73d98a0c51fdd143/python/bridgestan/compile.py#L124

File: rust/src/bs_safe.rs
Comment (randommm) [reply rust/src/bs_safe.rs:106]: @aseyboldt could you share your opinion on this?

File: rust/Cargo.toml
Comment (randommm) [reply rust/Cargo.toml]: well spotted, thanks! I've read this part of the Cargo book more than once but forgot about the implicit feature creation

**Comment by WardBrian** [comment]
> We could just have a global Mutex that protects the compilation, but that would still fail if multiple independent processes try to compile models in parallel? (@WardBrian Is this something that has been looked into on the stan side?)

This is an unfortunate consequence of Stan really only supporting in-tree builds. This has been frustrating for [other reasons](https://github.com/stan-dev/cmdstan/issues/1175), I don't think this specific one has come up though. 

The right thing to do if concurrent builds are requested is to do them all in the same call to `make` and let `make`'s jobserver do it, but that seems very difficult to expose in a meaningful way. I guess the compile function could take a list of filenames, but that also seems cumbersome.

**Comment by randommm** [comment]
> * I guess this is why you had to disable the test, or disable testing in parallel

Not exactly, it was just that i needed to delete the .so file before asking the function to compile the model, otherwise the function would skip compilation as the file already exists.

The problem then arises because other tests expect the .so file to exist in the meantime (they don't try to compile it first), so they will fail if they started after the file was deleted.

However, i think that if two processes tried to compile at the same time, i imagine that at least on Linux these wouldn't be a problem (on Windows, idk, conflict of two process opening the same file?).

**Comment by randommm** [comment]
> > We could just have a global Mutex that protects the compilation, but that would still fail if multiple independent processes try to compile models in parallel? (@WardBrian Is this something that has been looked into on the stan side?)
> 
> This is an unfortunate consequence of Stan really only supporting in-tree builds. This has been frustrating for [other reasons](https://github.com/stan-dev/cmdstan/issues/1175), I don't think this specific one has come up though.
> 
> The right thing to do if concurrent builds are requested is to do them all in the same call to `make` and let `make`'s jobserver do it, but that seems very difficult to expose in a meaningful way. I guess the compile function could take a list of filenames, but that also seems cumbersome.

so right now, if two processes call make at the same time, bad things happen? I haven't tried but imagined that it wouldn't be bad on Linux

**Comment by WardBrian** [comment]
I think it would be fine after the initial build? But if nothing has been successfully built yet I think you could definitely get a filesystem race on things like `main.o` that are built once and then re-used

File: rust/src/download_compile.rs
Comment (randommm) [reply rust/src/download_compile.rs:141]: okay, see 86e9ee8e2997caa4b8cd713b2ecb54bc6ce0ca75

File: .github/workflows/main.yaml
Comment (randommm) [reply .github/workflows/main.yaml:294]: done

**Comment by randommm** [comment]
>   Maybe we also want to check the downloaded stan version somehow? If we tie it to a specific release we could just use a checksum I guess? Although I'm actually not sure how we'd best do that given that the file we want to download contains the repo, which would have to include the checksum...

Check commit fab21a7c2a6b0355 for a possible solution. The disadvantage is that will be download an outdated version of bridgestan, so not sure if this can cause a chicken and egg issue in the future in case ABI changes edge cases.

File: rust/src/bs_safe.rs
Comment (aseyboldt) [reply rust/src/bs_safe.rs:106]: I'm not sure either. I guess I'd rather not gate it, and have the variant in the enum all the time.

File: rust/src/download_compile.rs
Comment (aseyboldt) [reply rust/src/download_compile.rs:88]: Oh no...
Maybe we could add a comment and mention that we can switch to https://doc.rust-lang.org/std/path/fn.absolute.html when that is stable?

File: rust/src/download_compile.rs
Comment (WardBrian) [reply rust/src/download_compile.rs:105]: A lot of Stan programs in the wild want to be able to do includes relative to their current directory, which requires this extra flag. Most other interfaces to Stan pass it by default, so it's done here to match user expectations. 

Depending on how you invoke `make` you need to make sure the path you're including is the directory the model is in, not the bridgestan folder.

**Comment by WardBrian** [comment]
> The disadvantage is that will be download an outdated version of bridgestan, so not sure if this can cause a chicken and egg issue in the future in case ABI changes edge cases.

I think this would definitely cause problems. Even lacking a major version bump, this would break even during minor version bumps if the new feature required new functions in the C++ layer.

You could instead do what we do for Julia:
We store the SHA of the [current version of the source,](https://github.com/roualdes/bridgestan/blob/main/julia/Artifacts.toml) before updating the Julia code itself in the release action
https://github.com/roualdes/bridgestan/blob/b41f83b61571d4744ba5521a73d98a0c51fdd143/.github/workflows/release.yaml#L61-L65

We could also start creating a tarball that is _only_ the C++ (this is probably a good idea for several reasons) which would make this chicken-egg problem easier

File: rust/src/download_compile.rs
Comment (aseyboldt) [reply rust/src/download_compile.rs:105]: Makes sense. I think we need to change it then so that we use the original dir.

File: rust/src/download_compile.rs
Comment (randommm) [reply rust/src/download_compile.rs:92]: okay, added the comments

File: rust/src/download_compile.rs
Comment (randommm) [reply rust/src/download_compile.rs]: should be fixed now

**Comment by randommm** [comment]
finally managed to fix the build on macos now!

**Comment by randommm** [comment]
@WardBrian @aseyboldt made the API a little more clear now regarding that bridgestan is being download, and also easier to point to a custom bridgestan dir. I think that the checksum and the lock dir would be better fit as others issues/PRs.

File: rust/Cargo.toml
Comment (WardBrian) [inline rust/Cargo.toml:22]: Thanks for the changes. I'll let @aseyboldt comment on the ones he suggested, but this is looking pretty good from my end. 

I've given this some more thought and realized that almost all of these dependencies are not specific to _compilation_, but rather they're all for the "download the source if you don't have it" feature (right?).

Would it be reasonable to have _that_ be what is gated behind a feature? If the crate wasn't compiled with that and can't find the sources, it could unconditionally return the DownloadFailed error with a message saying so.

Comment (randommm) [reply rust/Cargo.toml]: this is a good point, also helps on the security matter mentioned by @aseyboldt

File: rust/README.md
Comment (WardBrian) [inline rust/README.md:49]: Could we do something similar to the other examples and have this be a relative path (with the download option commented out as an example?)?

E.g.:
https://github.com/roualdes/bridgestan/blob/ef1f98694b9c6236e2d0b96eba025e864150e2de/julia/example.jl#L5-L9

File: rust/examples/example.rs
Comment (WardBrian) [inline rust/examples/example.rs:19]: Same as above: relative path pointing to the repo root?

File: rust/tests/model.rs
Comment (WardBrian) [inline rust/tests/model.rs:68]: To more similarly match our other tests on other platforms, could we use the `$BRIDGESTAN` environment variable? This is particularly useful because it lets us test `VERSION+1`, where as this test assumes always that the compilation code will also work with the currently released version.

It also then does not need to rely on the download feature.

File: rust/tests/model.rs
Comment (WardBrian) [inline rust/tests/model.rs:98]: Could this test be a bit more basic and just test that e.g. the path returned by downloading has a subfolder called `stan` and a file called `makefile`? I don't think there is anything necessary about testing compilation alongside the download, independent of the above test of just compilation?

File: rust/src/download.rs
Comment (WardBrian) [inline rust/src/download.rs:14]: [optional, if not in this PR I will circle back around to do it eventually] It would be nice to have a similar warning to what we do in Python if a lot of these exist:

https://github.com/roualdes/bridgestan/blob/ef1f98694b9c6236e2d0b96eba025e864150e2de/python/bridgestan/compile.py#L73-L78

File: rust/src/compile.rs
Comment (WardBrian) [inline rust/src/compile.rs:14]: Question as much for @aseyboldt as @randommm: The `AsRef<Path>` generic is the canonical example I have seen for the [Non-Generic Inner Function](https://www.possiblerust.com/pattern/non-generic-inner-functions) pattern: do we care about doing something like that here? I have no sense of if this is useful outside things like stdlib functions you expected to get called in many different ways

**Review by WardBrian — COMMENTED** [review]
The following are hopefully my final comments, focusing on some edge cases in the example and tests. 

I'd still like @aseyboldt do give another look over for any Rust nits I'm missing. But I think this is very good!

File: rust/src/compile.rs
Comment (randommm) [reply rust/src/compile.rs]: fair point, idk, it was just a convenience to the user, but i think it's not worth the need for monomorphisations and such, so i've removed it.

File: rust/tests/model.rs
Comment (WardBrian) [inline rust/tests/model.rs:84]: This should be platform independent

File: rust/tests/model.rs
Comment (WardBrian) [inline rust/tests/model.rs:50]: Does this still require unix-only when it is run on its own due to the `[ignore]`?

**Review by WardBrian — APPROVED** [review]
This looks good to me. I left two tiny questions about the tests, and I would still appreciate if @aseyboldt could have a look before merging.

Thanks again!

**Pull request merged**
